Fixes text decoding error when Tf command appears before BT. (#542)

cris-shi opened the pull request:
### Issue Description
The [#Issue 542](https://github.com/smalot/pdfparser/issues/542) caused by the Tf command appears before the text block. And the method PDFObject::getSectionText only extracts command within BT & ET.

```pdf
Q
0.18 Tc
/F1 3.05 Tf
BT
0 0 0 sc
5.66929 7.65354 Td
<007D00F0010900C5012B00D0000300F001090128012B00F0010900D10003012E0137012B0003010300AD0128010F012E013400D0021800E5012B> Tj
ET
```

### Solution
```php
public function getSectionsText(?string $content): array
{
    ...
    if (preg_match_all('/(\sQ)?\s+(.*?)BT[\s|\(|\[]+(.*?)\s*ET(\sq)?/s', $textCleaned, $matches, \PREG_OFFSET_CAPTURE)) {
        foreach ($matches[3] as $pos => $part) {
            ...

            // Add Tx commands which before BT.
            // @see: https://github.com/smalot/pdfparser/issues/542
            if (!empty($matches[2][$pos][0]) && preg_match('/\sTf\s/', $matches[2][$pos][0])) {
                $section = trim($matches[2][$pos][0].$section);
            }

            ...
        }
    }
    ...
}
```

k00ni reviewed (changes requested):
@cris-shi first of all, thank you for your pull request!

* [ ] Can you add 1 or 2 PHPUnit tests which show that your changes work?

Adding these if-clauses might solve the issue for now, but it also adds complexity. Developer now have to take these also into account when working with the code. 
* [ ] An improvement would be to add further explanation/documentation. Why was it changed? What didn't work before? What is possible from now on?
* [ ] Another point is the position of this fixes. Is this the only way to handle these flags?
* [ ] Last point: Please split up these one-liners to increase readability. You are not only making a lot of assumption (which may fail during runtime), but also stack conditions with function calls.

I am looking forward on your feedback and getting this to work.

cris-shi commented:
And I did a little more research. From what I understand, these text parsing errors are caused by not selecting the correct text font, which contains CMap table.

Outside the text block, there are three commands that will change the current font:
Q: Restore graphics state
q: Save graphics state
Tf: Set text font and size

So I rewrote the code in the function PDFObject::getSectionsText to extract all the commands of Q, q, Tf before the text block, which might affect the font selection of the text block.

> 
> @cris-shi first of all, thank you for your pull request!
> 
> * [ ]  Can you add 1 or 2 PHPUnit tests which show that your changes work?
> 
> Adding these if-clauses might solve the issue for now, but it also adds complexity. Developers now have to take these also into account when working with the code.

I added a PHPUnit test with pdf sample provided by issue 542.

> * [ ]  An improvement would be to add further explanation/documentation. Why was it changed? What didn't work before? What is possible from now on?

In most cases, the font is defined in the text block, the original code work fine with this.
But in this case, the font is defined before the text block.

> * [ ]  Another point is the position of this fixes. Is this the only way to handle these flags? Architecture of PDFParser isn't the best and it grew organically.

Handling these flags worked fine. The problem is extracting these flags.
As I know, this is the only position to extract text blocks, so I think this fixes is enough.

> * [ ]  Last point: Please split up these one-liners to increase readability. You are not only making a lot of assumptions (which may fail during runtime), but also stack conditions with function calls.
> 
> I am looking forward to your feedback and getting this to work.

k00ni reviewed (changes requested):
Hey! Thank you for taking the time.

You wrote: 
> And I did a little more research. From what I understand, these text parsing errors are caused by not selecting the correct text font, which contains CMap table.

Can you add a link (or part reference) to your source?

* [ ] I appreciate the new test, but please revert your changes in `testGetSectionText`. Please copy it and make your additions. This way we preserve the meaning of the old test and test PDFParser with similar parameters. Same goes for `testGetSectionText` of PageTest.

k00ni commented on src/Smalot/PdfParser/PDFObject.php:
```diff
@@ -191,7 +191,7 @@ public function getSectionsText(?string $content): array
 
         // Extract text blocks.
-        if (preg_match_all('/(\sQ)?\s+BT[\s|\(|\[]+(.*?)\s*ET(\sq)?/s', $textCleaned, $matches, \PREG_OFFSET_CAPTURE)) {
+        if (preg_match_all('/(.*?)\s+BT[\s|\(|\[]+(.*?)\s*ET(?=\s|$)?/s', $textCleaned, $matches, \PREG_OFFSET_CAPTURE)) {
```
```suggestion
        $regex = '/(.*?)\s+BT[\s|\(|\[]+(.*?)\s*ET(?=\s|$)?/s';
        if (preg_match_all($regex, $textCleaned, $matches, \PREG_OFFSET_CAPTURE)) {
```

Makes it easier to read.

k00ni commented on src/Smalot/PdfParser/PDFObject.php:
```diff
@@ -184,14 +184,49 @@ public function cleanContent(string $content, string $char = 'X')
+     */
+    private function addQAndqFlagsAndTfCommands(string $section, $matches, int $pos): string
+    {
+        if (preg_match_all('/(?:\s|^)([Qq])(?:\s|$)/', $matches[1][$pos][0], $qMatches, \PREG_OFFSET_CAPTURE)) {
+            $len = \strlen($matches[1][$pos][0]);
+            $matchesCount = count($qMatches[0]);
+            for ($i = $matchesCount - 1; $i >= 0; --$i) {
+                $str = substr($matches[1][$pos][0], $qMatches[0][$i][1] + 3, $len - ($qMatches[0][$i][1] + 3));
+                $len = $qMatches[0][$i][1];
+                if (preg_match('/\sTf(\s|$)/', $str)) {
+                    $section = trim($str)."\n".$section;
+                }
+
+                if ('Q' == $qMatches[1][$i][0]) {
+                    $section = "Q\n".$section;
+                } elseif ('q' == $qMatches[1][$i][0]) {
+                    $section = "q\n".$section;
+                }
+            }
+            $str = substr($matches[1][$pos][0], 0, $qMatches[0][0][1]);
+            if (preg_match('/\sTf(\s|$)/', $str)) {
+                $section = trim($str)."\n".$section;
+            }
+        } elseif (preg_match('/\sTf(\s|$)/', $matches[1][$pos][0])) {
+            $section = trim($matches[1][$pos][0])."\n".$section;
+        }
+
+        return $section;
```
I moved this to a separate function. 

If possible, add the comments you made here also in code. That would bolster code readability and understanding.

k00ni commented:
@cris-shi What is the status here? Feel free to tell me, if you need assistance.

k00ni commented:
Closed for now, please reopen when you want to work on it again.
